test: enable multiprocess tests on Win, Mac and Linux

zain-mecklai opened the pull request:
This first PR will make some basic changes that will eventually allow cross platform testing for multiprocess tests.

zain-mecklai committed:
Start the process of getting multiprocess tests working on Windows and Mac

zain-mecklai committed:
test: enable multiprocess tests on Windows and mac

NoelStephensUnity reviewed (approved):
Looks like a good start.

0xFA11 commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -43,10 +66,13 @@ public static void StartWorkerNode()
             throw;
         }
 
+        string logPath = Path.Combine(s_MultiprocessDirInfo.FullName, $"zlogfile{Processes.Count}");
```
intentional `(z)logfile` prefix in the name?

zain-mecklai replied:
I just wanted to make it unique enough to be easily seen. I can call it logfile or whatever seems best.

0xFA11 replied:
oh, I see. perhaps `$"logfile-mp{Processes.Count}"`?

zain-mecklai replied:
done

0xFA11 commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -34,9 +34,9 @@ public abstract class BaseMultiprocessTests
         [OneTimeSetUp]
         public virtual void SetupTestSuite()
         {
-            if (ShouldIgnoreTests)
+            if (IsPerformanceTest)
```
much clear and obvious, thank you for this!

0xFA11 commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -43,10 +66,13 @@ public static void StartWorkerNode()
         workerProcess.StartInfo.RedirectStandardError = true;
         workerProcess.StartInfo.RedirectStandardOutput = true;
-        workerProcess.StartInfo.Arguments = $"{IsWorkerArg} -popupwindow -screen-width 100 -screen-height 100";
+        workerProcess.StartInfo.Arguments = $"{IsWorkerArg} -popupwindow -screen-width 100 -screen-height 100 -logFile {logPath}"; // Let each process log to a different file
```
should we have popup windows or run these processes in -batchmode instead?

(perhaps not a great topic for this discussion but wanted to put it here anyway)

zain-mecklai replied:
I'm not sure, this is how it was constructed before. I didn't want to change what I didn't understand.

0xFA11 replied:
fair enough but I'd like to summon @SamuelBellomo here as he's the original author :)

SamuelBellomo replied:
good point, these arguments were added with local testing in mind, they should be adapted to headless when running in CI

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Add Debugging to find who or what is null

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/SnapshotSystem.cs:
```diff
@@ -651,6 +651,15 @@ public void Dispose()
 
         public void NetworkUpdate(NetworkUpdateStage updateStage)
         {
+            if (m_NetworkManager == null)
+            {
+                Debug.Log("m_NetworkManager is null");
+            }
+            else if (m_NetworkManager.NetworkConfig == null)
+            {
+                Debug.Log("m_NetworkManager.NetworkConfig is null");
+            }
```
I assume this is for testing/debugging but not for actual PR?

zain-mecklai replied:
Oh darn, yes, let me remove that.

zain-mecklai committed:
Build multiprocess tests on Windows and Mac

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Revert change to SnapshotSystem

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

SamuelBellomo reviewed (approved):
beside the argument comment, lgtm

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

0xFA11 commented on .yamato/project-tests.yml:
```diff
@@ -41,6 +41,9 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
     - npm install upm-ci-utils@stable -g --registry https://artifactory.prd.cds.internal.unity3d.com/artifactory/api/npm/upm-npm
+    - pip install unity-downloader-cli --index-url https://artifactory.prd.it.unity3d.com/artifactory/api/pypi/pypi/simple
+    - unity-downloader-cli -u {{ editor }} -c editor -w --fast
+    - {{ platform.editorpath }} -projectpath testproject -batchmode -quit -logfile BuildMultiprocessTestPlayer.log -executeMethod Unity.Netcode.MultiprocessRuntimeTests.BuildMultiprocessTestPlayer.BuildDebug 
```
so, this is multiprocess testing in the same box/machine, right?

zain-mecklai replied:
I tried initially to do everything in one large PR earlier but it was too complicated and too many lines so I'm taking it one step at a time.

0xFA11 replied:
no, I would actually prefer individual PRs going through :) I was just double-checking my understanding.

zain-mecklai committed:
Remove tabs in document and replace with spaces

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Enable Linux Editor/Player tests for Multiprocess

NoelStephensUnity committed:
fix

fixes issues with clients being out of alignment with the server's scene in build list.
Includes client process command line argument to load the MultiprocessTestScene.
fixes issues with release MultiProcess client builds
fixes missing PrefabReference script component in MultiprocessTestScene's NetworkManager.

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity reviewed (approved):
Great start Zain!
Looks good to me.

0xFA11 reviewed (approved):
(ship-it-parrot)

NoelStephensUnity committed:
fix

We should destroy the GameObject and not the component to avoid dependency issues.

NoelStephensUnity committed:
fix

MultiProcess tests no longer unload the TestRunner scene.
The server side MultiProcess scene is loaded additively and then set as the active scene until the OneTimeTearDown attribute decorated method TeardownSuite is invoked.
Added VerifySceneBeforeLoading callback handler to server side NetworkSceneManager in order to avoid sending the TestRunner scene as a scene that needs to be "synchronized" during client synchronization.
Adjusted the FixedUpdateMessagesAreOnlyProcessedOnceTest to not timeout so quickly as it was failing due to timing out too early.  This could have something to do with the last multiprocess client hanging around for 20 seconds before unloading.

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity committed:
test-fix

Testing to see if just killing the processes at the end of a test suite will fix the issue where one client works but two clients fails.

NoelStephensUnity committed:
test-fix

Trying to determine where some of the issues are occurring when running in Yamato.
Adjusting when we create the processes as well as added additional debug information in the event that we run into the scenario where there is more than one TestCoordinator.

NoelStephensUnity committed:
test-fix

Trying this out to see how TestFixtures might be getting executed differently in Yamato than locally?  Several checks to see if we have already loaded the MultiProcessTests scene.

0xFA11 commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -31,23 +31,40 @@ public abstract class BaseMultiprocessTests
         private void OnSceneLoaded(Scene scene, LoadSceneMode mode)
         {
             SceneManager.sceneLoaded -= OnSceneLoaded;
+            if(scene.name == BuildMultiprocessTestPlayer.MainSceneName)
```
`./standards.py --fix` :)

NoelStephensUnity replied:
@MFatihMAR 
Yeah... ignore all of the next series of "test-fix" commits.
Evidently, it appears that OneTimeSetUp runs on a per TestFixture pass =or= scenes are not unloading before the next TestFixture instance is started...what I have determined is that there is something about the "OneTimeSetUp" that is handled differently in Yamato than if you run TestRunner in the editor...thus why everything passes in the TestRunner editor locally but has been failing in Yamato. (assuming "OneTimeSetUp" could run for each TestFixture seems to be getting me further along in the MultiProcess tests with the last commit passing on all of the ExecuteStepInContextTests)  

So, anything with "test-fix" in the commit is just me trying to debug/fix this difference...I will make sure when I am done to do a full standards check... and we should probably have everyone "re-review" the changes made as there will definitely be some differences.

NoelStephensUnity committed:
test-fix

Excluding the NetworkVariablePerformanceTests for this run.
It has similar issues to the ExecuteStepInContext tests with the OneTimeSetup.

NoelStephensUnity committed:
test-fix

Removing the last TestCoordinatorTests TestFixture to determine if the MultiProcessTestScene is still hanging around when all MultiProcess tests are done or if it is related to the failure of the last 2 client TestCoordinatorTests TestFixture.

zain-mecklai committed:
Simplify logic on when to run teardown for multiprocess tests

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Revert changes

zain-mecklai committed:
Merge

NoelStephensUnity committed:
fix

This is the "bare minimum" adjustments to getting the multiprocess code base back to where it was running but having issues with TestFixtures.

zain-mecklai committed:
Enable Multiprocess tests

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Restore tests and try again

zain-mecklai committed:
codesign

zain-mecklai committed:
Fix job definition file so that it is valid

zain-mecklai committed:
disable graphics for the multiprocess tests

zain-mecklai committed:
Different args required for different platforms

zain-mecklai committed:
fix formatting

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Fix conditional syntax to validate yaml

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -72,10 +87,13 @@ public static void StartWorkerNode()
         workerProcess.StartInfo.RedirectStandardOutput = true;
-        workerProcess.StartInfo.Arguments = $"{IsWorkerArg} -popupwindow -screen-width 100 -screen-height 100 -logFile {logPath}"; // Let each process log to a different file
+        // In order to support interactive mode there will need to be a way to differentiate the various operational modes and turn graphics on and off based on it
+        // workerProcess.StartInfo.Arguments = $"{IsWorkerArg} -popupwindow -screen-width 100 -screen-height 100 -logFile {logPath} -s {BuildMultiprocessTestPlayer.MainSceneName}";
```
with the extra args above, is this comment still needed?

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -89,9 +107,15 @@ public static void StartWorkerNode()
+        foreach (var process in Processes)
+        {
+            Debug.Log($"Killing process {process.Id} with state {process.HasExited}");
+            process.Kill();
```
from what I read here
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill?view=net-5.0
kill will throw 
```
InvalidOperationException
The process has already exited.
```
if the process is already dead, I'm guessing we still want to handle that case?

zain-mecklai replied:
Good point, I'll handle this better.

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/NetworkVariablePerformanceTests.cs:
```diff
@@ -1,3 +1,4 @@
+#if ONE_TIME_SETUP_WIP
```
what is this?

SamuelBellomo replied:
with the isPerf check, do we need this?

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -22,7 +22,7 @@ public class TestCoordinator : NetworkBehaviour
     public const int PerTestTimeoutSec = 5 * 60; // seconds
 
-    public const float MaxWaitTimeoutSec = 20;
+    public const float MaxWaitTimeoutSec = 30;
```
I'd be down to have different timeout values depending on the case. Having everything depend on the same val for all timeouts might not be the best. this timeout is used in the tests themselves, 30 seconds for inter process communication timeout is pretty long

zain-mecklai replied:
reverting

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -101,12 +118,21 @@ public void TestRunTeardown()
         m_TestResultsLocal.Clear();
     }
 
+    private void OnDisable()
```
OnDisable or OnDestroy? Cause OnDisable usually means this same instance can be reenabled later.

SamuelBellomo replied:
If we do disable, then enable, we'd be in an invalid state no?

zain-mecklai replied:
@NoelStephensUnity can you speak to these changes as I believe you made them?

zain-mecklai replied:
It looks like this is done in OnDestroy as well. I'll ping @NoelStephensUnity about this change.

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -254,7 +280,7 @@ public void TriggerActionIdClientRpc(string actionId, byte[] args, ClientRpcPara
         {
             WriteErrorServerRpc(e.Message);
-            throw;
+            //throw;
```
?

zain-mecklai replied:
Yup, this is weird, reverting.

zain-mecklai committed:
Revert changes as it isn't clear we need them

NoelStephensUnity committed:
Refactor

The base requirements to build and run multiprocess stuff.

zain-mecklai committed:
filter out the mobile platforms as the test doesn't run on them directly

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity committed:
style

removing LF

NoelStephensUnity committed:
fix

Adding back the multiprocess ignore capabilities with the ability to easily override this ignore for continued development work.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiprocess-fixes

NoelStephensUnity committed:
fix

including the IgnorMultiprocessTests test check for the NetworkVariablePerformanceTests

NoelStephensUnity committed:
style

fixing white space formatting issue due to indentation.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiprocess-fixes

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Merge

zain-mecklai committed:
merge with develop

zain-mecklai committed:
Revert change back to original value

zain-mecklai committed:
Set display for ubuntu since we aren't using centos anymore

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac
